torchao/quantization/subclass.py

```diff
@@ -592,3 +596,225 @@ def to_qtensor_components(cls, input_float, groupsize=128, inner_k_tiles=8):
+        int_data = quantize_affine(input_float, block_size, scale, zero_point, target_dtype, quant_min, quant_max)
+        return cls(
+            int_data,
+            scale,
+            zero_point,
+            quant_min,
+            quant_max,
+            block_size,
+            input_float.shape,
+            input_quant_func=input_quant_func,
+            dtype=input_float.dtype
+        )
+
+    # __torch_function__ = torch._C._disabled_torch_function_impl
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        kwargs = {} if kwargs is None else kwargs
+
+        if func is torch.nn.functional.linear:
+            input_tensor, weight_qtensor, bias = (
+                args[0],
+                args[1],
+                args[2] if len(args) > 2 else None,
+            )
+            if weight_qtensor.input_quant_func is not None:
+                input_tensor = weight_qtensor.input_quant_func(input_tensor)
+                input_tensor = input_tensor.dequantize()
+            weight_tensor = weight_qtensor.dequantize()
+            return torch.nn.functional.linear(input_tensor, weight_tensor, bias)
+
+        try:
+            with torch._C.DisableTorchFunctionSubclass():
+                return func(*args, **kwargs)
+        except:
+            print(f"ERR: subclass doesn't implement {func}")
+
+
+    def _get_to_kwargs(self, *args, **kwargs):
+        device, dtype, _, memory_format = torch._C._nn._parse_to(*args, **kwargs)
+        device = self.device if device is None else device
+        dtype = self.dtype if dtype is None else dtype
+        memory_format = (
+            memory_format if memory_format is not None else torch.preserve_format
+        )
+        kwargs = {
+            "device": device,
+            "dtype": dtype,
+            "memory_format": memory_format,
+        }
+        return kwargs
+
+    def to(self, *args, **kwargs):
+        kwargs = self._get_to_kwargs(*args, **kwargs)
+        return self.__class__(
+            self.int_data.to(kwargs["device"]),
+            self.scale.to(kwargs["device"]),
+            self.zero_point.to(kwargs["device"]),
+            self.quant_min,
+            self.quant_max,
+            self.block_size,
+            self.shape,
+            self.input_quant_func,
+            **kwargs,
+        )
+
+    def _apply_fn_to_data(self, fn):
+        return self.__class__(
+            fn(self.int_data),
+            fn(self.scale),
+            fn(self.zero_point),
+            self.quant_min,
+            self.quant_max,
+            self.block_size,
+            self.shape,
+            self.input_quant_func,
+            dtype=self.dtype,
+        )
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        # two scenarios where we currently fall back to vanilla mm:
+        # 1 - when tensor is on CPU: we are missing qmm for CPU, but we should have a CPU implementation
```

Comment:
I only see `is_cuda` version, did you mean to add cpu when we don't have qmm?

Reply:
so right now the version is just for 8da4w + executorch. we actually haven't added cpu or cuda version. in the future:

1. we'll add cpu/cuda version (int4mm etc.)
2. we'll need to hide the 8da4w executorch version under some options like layouts

Reply:
oh I will need to change is_cuda to is_cpu here actually

Reply:
> cpu device + et laytout --> gives current 8da4w executorch representation
> cpu device + avx layout --> gives optimized kernel for 8da4w in avx cpu etc

Barring some quant parameters, isn't the weight representation kernel independent i.e. weight elements will be int8 with values in [qmin, qmax] range?

The actual packed weight layout can be quite different and may not be able to fit in the `tensor` representation

Reply:
to have efficient executorch in eager (e.g. for cpu or cuda), we would like to produce packed weight directly in this quantized Tensor I think.

but int8 with qmin/qmax could be an intermediate "unpacked" format that can serve as a standard representation that we can use to translate between different packing format, e.g. 
1. first the tensor is packed for some cpu efficient kernel
2. then user calls tensor.to("cuda") (or model.to("cuda")), we can first unpack the tensor to int8 + qmin/qmax representation, and then repack to the cuda format

Reply:
I am not sure what we are trying to achieve here with layout. We are mixing execution semantics with the semantics of quantization. If we do this, we will keep adding n different layouts to this class. Plus in some cases ahead of time packing is not possible since your optimal execution may depend on the machine your the code. 

In PT2E we explicitly moved away from quantized tensor ((and we can argue whether it is the right choice and would like to see why not) and here we are going in the opposite direction of not only having a quantized tensor but including execution semantics in it.

Reply:
I see, yeah I think that makes sense, I believe the main reason we are mixing these here is probably because that's how the rest of pytorch tensor and operators are structured. But I'll write down more details and we can discuss further

Reply:
@kimishpatel - I don't quite follow, why is using a layout for a particular bit packing structure not the right abstraction?

Sure, some bit packing layouts are currently very useful to certain accelerators, but perhaps some of them translate to others? Maybe there's some similarities between them? I think it's also still ongoing research which one is the best (or at least it seems like it is).

Reply:
> **@kimishpatel - I don't quite follow, why is using a layout for a particular bit packing structure not the right abstraction?

Because, we will continue to introduce execution dependent packing layouts as part of this tensor subclass. e.g. 5 year older x86/ARM want to pack int4/int8 tensor with A, newer ones may want it in way B and future ones want in way C (not to say about different layouts that different classes of GPUs may want). So this class assuming the responsibility of encoding all the different packing layouts does not seem right to me.

I can be convinced otherwise, but it feel that it would be cleaner if there was a canonical representation that this class assumes and then it can be either extended, for custom layouts/packing, or introduce custom ops that do custom packing.

Reply:
@kimishpatel - yes, but that's ok though, no? I mean, wouldn't you want to pick the packing layout that's best for the available CPU? Also, we give these layouts names. You can still allocate a GPU specific layout on your raspberry PI. It just might not run (at least not efficiently) locally.

Reply:
The bitpacking layout seems independent from the allocator used to allocate the memory. You can use malloc to allocate a bitpacking layout that's optimized for CPU or you can use cuda malloc to allocate a bitpacking layout that's optimized for the Skylake CPU. But maybe this is an overly naive view?

Reply:
Personally I feel it is trying to do too much because the layout you want is maybe tied to kernel you wrote. Then say the output of your op is also AQT which can also have a special layout. Which means every op that you register for AQT must handle all the packed layouts. Like in another comment you had an example of `gelu` op being registered for `nf4`. Now if `nf4` tensor had multiple packings/layouts then the registered op will have to interpret each of them. (Maybe you need a canonical representation where to/from tranforms are possible)

So the question is do you allow for arbitrary packing in AQT via some attribute + to/from tranformations for canonical representation OR
decouple it in a way that any special consideration like packing is handled elsewhere.

And note that this has to play well with export and other systems as well, else we are just solving eager mode problem.